src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopies = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    let nextState;
+
+    switch (action.type) {
+      case 'addProperties':
+        nextState = { ...stateCopy, ...action.extraData };
+        break;
+      case 'removeProperties':
+        nextState = { ...stateCopy };
+
+        for (const key of action.keysToRemove) {
+          if (nextState.hasOwnProperty(key)) {
+            delete nextState[key];
+          }
+        }
+        break;
+      case 'clear':
+        nextState = {};
+        break;
+    }
+    stateCopies.push(nextState);
+    stateCopy = nextState;
```

Comment:
Instead of assigning `nextState` to `stateCopy` inside the loop, consider using `stateCopy` directly. This way, you avoid having an unnecessary variable and make the code cleaner. After each case in the switch, you can directly push `stateCopy` into `stateCopies`.